### DISCUSSION TITLE
Best preset settings for Pony model?

### DISCUSSION THREAD

**Discussion by mashb1t** [lllyasviel/Fooocus#3217] [Q&A]
Support for VAE download has been implemented in https://github.com/lllyasviel/Fooocus/pull/3216.
Proposed preset: https://github.com/lllyasviel/Fooocus/blob/bbf90a97b7ebd1808836fb39d9e490966070f4c2/presets/pony_v6.json
Inpainting doesn't work with any inpaint engine other than "None", same as for the playground model.

Happy to hear your insights about CFG (recommended: 5-7, proposal: 7) and positive + negative prompts!

Examples (fantasy themed & unmodified, metadata in images):

| CFG (Guidance Scale) | Tiefling Shaman | Mouse Wizard | Bear Druid |
|--------|--------|--------|--------|
| 7 | ![image](https://github.com/lllyasviel/Fooocus/assets/9307310/0acf3ff4-d066-4e6b-aab2-ae52a4324917) | ![image - 2024-07-01T132556 062](https://github.com/lllyasviel/Fooocus/assets/9307310/a1704d04-c87b-4f80-a9d3-b1d152a2671b) | ![image](https://github.com/lllyasviel/Fooocus/assets/9307310/6897c884-a423-48bb-9c39-dbfe8c7fc400) | 
| 6 | ![image - 2024-07-01T131816 664](https://github.com/lllyasviel/Fooocus/assets/9307310/95be157c-356e-4fe8-981f-17e88e687425) | ![image - 2024-07-01T124036 929](https://github.com/lllyasviel/Fooocus/assets/9307310/0113c2c5-c94d-4b30-9f9e-dbfb29b7a45b) | ![image - 2024-07-01T132228 000](https://github.com/lllyasviel/Fooocus/assets/9307310/dd7db368-039e-4fd9-8192-1a88e0ec120c) |
| 5 | ![image - 2024-07-01T131857 813](https://github.com/lllyasviel/Fooocus/assets/9307310/94fecd6c-d442-465a-b804-cdfec64bd387) | ![image - 2024-07-01T124106 725](https://github.com/lllyasviel/Fooocus/assets/9307310/06389478-c706-40f7-af5c-fa933436fd40) | ![image - 2024-07-01T132255 765](https://github.com/lllyasviel/Fooocus/assets/9307310/3a3d9573-ae03-4f6d-b724-042cde55818f) |

**Comment by ILikeToasters** [comment]
That actually is pretty close to what I have narrowed things down to. Although I am still experimenting with sampler but will say that one works well for the realistic ones I have tested for sure. Only difference I have done is creating a style instead of the words in the prompt.

**Reply by mashb1t** [reply]
@ILikeToasters thanks, already adjusted and it's now a style. Also included in the pony feature PR are adjustments to change the inpaint mode to improve details on preset change for pony and playground.
=> preventing accidental wrong results due to incompatible inpaint engine.

**Reply by ILikeToasters** [reply]
As for the inpainting and setting it to improve details, that will end up also changing the other setting like denoising strength and respective field. Not sure that makes sense. Would it not be better to just switch the inpaint engine to none? Also I just switched over to the mashb1t:feature/add-pony-support-and-preset branch and got an error: 

```
UNet ADM Dimension 2816
Traceback (most recent call last):
  File "F:\mashb1t test\Fooocus\modules\patch.py", line 465, in loader
    result = original_loader(*args, **kwargs)
  File "F:\mashb1t test\python_embeded\lib\site-packages\safetensors\torch.py", line 259, in load_file
    with safe_open(filename, framework="pt", device=device) as f:
FileNotFoundError: No such file or directory: "F:\\mashb1t test\\Fooocus\\models\\vae\\ponyDiffusionV6XL_vae.safetensors"

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "F:\mashb1t test\Fooocus\modules\async_worker.py", line 998, in worker
    handler(task)
  File "F:\mashb1t test\python_embeded\lib\site-packages\torch\utils\_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "F:\mashb1t test\python_embeded\lib\site-packages\torch\utils\_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "F:\mashb1t test\Fooocus\modules\async_worker.py", line 469, in handler
    pipeline.refresh_everything(refiner_model_name=refiner_model_name, base_model_name=base_model_name,
  File "F:\mashb1t test\python_embeded\lib\site-packages\torch\utils\_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "F:\mashb1t test\python_embeded\lib\site-packages\torch\utils\_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "F:\mashb1t test\Fooocus\modules\default_pipeline.py", line 250, in refresh_everything
    refresh_base_model(base_model_name, vae_name)
  File "F:\mashb1t test\python_embeded\lib\site-packages\torch\utils\_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "F:\mashb1t test\python_embeded\lib\site-packages\torch\utils\_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "F:\mashb1t test\Fooocus\modules\default_pipeline.py", line 74, in refresh_base_model
    model_base = core.load_model(filename, vae_filename)
  File "F:\mashb1t test\python_embeded\lib\site-packages\torch\utils\_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "F:\mashb1t test\python_embeded\lib\site-packages\torch\utils\_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "F:\mashb1t test\Fooocus\modules\core.py", line 147, in load_model
    unet, clip, vae, vae_filename, clip_vision = load_checkpoint_guess_config(ckpt_filename, embedding_directory=path_embeddings,
  File "F:\mashb1t test\Fooocus\ldm_patched\modules\sd.py", line 470, in load_checkpoint_guess_config
    vae_sd = ldm_patched.modules.utils.load_torch_file(vae_filename_param)
  File "F:\mashb1t test\Fooocus\ldm_patched\modules\utils.py", line 13, in load_torch_file
    sd = safetensors.torch.load_file(ckpt, device=device.type)
  File "F:\mashb1t test\Fooocus\modules\patch.py", line 481, in loader
    raise ValueError(exp)
ValueError: No such file or directory: "F:\\mashb1t test\\Fooocus\\models\\vae\\ponyDiffusionV6XL_vae.safetensors"

Total time: 5.09 seconds
```

**Reply by mashb1t** [reply]
@ILikeToasters going to check the VAE download again.

> As for the inpainting and setting it to improve details, that will end up also changing the other setting like denoising strength and respective field. Not sure that makes sense. Would it not be better to just switch the inpaint engine to none?

Not really, as these settings are tailored to the corresponding inpainting mode and do work best like this as preset. I'll double check nevertheless.

**Reply by mashb1t** [reply]
@ILikeToasters fixed in https://github.com/lllyasviel/Fooocus/pull/3216/commits/c44f331cee4499522f57adcd2452684ddb53ee0e, where a key was missing for handling vae_downloads when loading a preset.

**Reply by ILikeToasters** [reply]
Works now. As for the inpainting, I just tested things out to make sure I was understanding what you meant and just tried with that branch and just inpainting without adjusting any settings. The issue is that if someone goes to do inpainting and expects that the masked area will be completely regenerated it won't happen since the denoising strength is set at 0.5.

**Reply by mashb1t** [reply]
@ILikeToasters i get that, but the alternative is to get messed up images as the inpaint models aren't compatible. I think this is the better way compared to leaving things as is, so uneducated users can just use the features without any issues. What's your proposal?

**Reply by ILikeToasters** [reply]
I am not sure of a perfect solution honestly but I know I have just been changing the inpaint engine to none. Obviously it doesn't work as well but it will still give me more expected results as opposed to using the improve details which honestly most people will probably think it is not doing anything since it doesn't change much. I find I can get decent result for some things if not masking a big area. It has been one of the big things I have been trying to figure out with pony models since I do lots of inpainting.

**Reply by ILikeToasters** [reply]
Honestly I have been doing inpainting a lot using non pony models.

**Reply by BoJangles002** [reply]
https://github.com/lllyasviel/Fooocus/discussions/3190#discussioncomment-9902837

Why isn't this the solution? Inpainting works flawless after changing the setting mentioned.

**Reply by mashb1t** [reply]
@ILikeToasters I mean, one can always switch again to other inpaint modes if necessary or can adjust the preset to your liking / create a new one. It's about users, who don't know anything about Fooocus, just switched the preset and then get weird results.
100% the new preset key `default_inpaint_method` will find its way to the playground preset, but pony is debatable. It does not work well but better than just random noise compared to playground.

@BoJangles002 thanks for the reference. The inpaint model isn't trained for pony, so without it you might get better results by default. Also setting overrides in presets practically disables refiners, which would be better then to not even show the refiner select.

**Reply by mashb1t** [reply]
@BoJangles002 (temporarily) pushed to PR for evaluation, see https://github.com/lllyasviel/Fooocus/pull/3216/commits/49d898f6c78ccec5a7ce132a36b08765fdb01e53

**Reply by mashb1t** [reply]
@ILikeToasters @BoJangles002 after running various tests for combinations of presets, models and overrides the result is to only override the inpaint engine instead of setting the whole inpaint mode (playground + pony) + use override refiner steps (pony).

**Reply by mashb1t** [reply]
done in [33c2292](https://github.com/lllyasviel/Fooocus/pull/3216/commits/33c22929e4f0965862e4957cc6ba4d99d4324515#diff-1a8fa8f0ed26b1f6ef903bbc6f09cdd307c51e27c231ccbcd9d72392509942ceL45-R44), please test again.

**Reply by ILikeToasters** [reply]
At the moment that seems to be the "best" results. Although I have not given up on testing different things yet since the best is still not great but especially when doing minor things it does decent I feel.

**Comment by BoJangles002** [comment]
I would run all those images again at cfg 5/6 and see if any of the details improve. specifically things like the tiefling horns that blend into the background, the trees/birds, and the bear's sash that morphs into his fur.

The only other thing I notice is that you have censored in the negative. Uncensored in the positive is supposedly significantly better but I don't understand the inner workings to know how/why. Something something unet something something dataset.

**Reply by mashb1t** [reply]
@BoJangles002 thanks for the feedback, will check again with different CFG.

Censored has already been removed from the negative prompt: https://github.com/lllyasviel/Fooocus/pull/3216/files#diff-f106c4c5729f1dc284aafa1254813175da0bca93c15870d94c8fc6d67931fdc6R37

**Reply by mashb1t** [reply]
@BoJangles002 you can find CFG 7/6/5 now[ in the table above](https://github.com/lllyasviel/Fooocus/discussions/3217#discussion-6879257).

**Comment by jaimeastorga2000** [comment]
The base Pony model is relatively low quality and unpredictable without a good style LoRA. Recommend either pick a default LoRA or switch to one of the AutismMix models for a default anime look.
